tests/tlsctx.c

```diff
@@ -33,6 +37,161 @@ static void ossl_err_print(void)
     }
 }
 
+static EVP_PKEY *load_key(const char *uri)
+{
+    OSSL_STORE_CTX *store;
+    OSSL_STORE_INFO *info;
+    EVP_PKEY *key = NULL;
+
+    if (!uri) {
+        fprintf(stderr, "Invalid NULL uri");
+        ossl_err_print();
+        exit(EXIT_FAILURE);
+    }
+
+    store = OSSL_STORE_open(uri, NULL, NULL, NULL, NULL);
+    if (store == NULL) {
+        fprintf(stderr, "Failed to open store: %s\n", uri);
+        ossl_err_print();
+        exit(EXIT_FAILURE);
+    }
+
+    if (strncmp(uri, "pkcs11:", 7) && strstr(uri, "type=private") == NULL) {
+        /* This is a workaround for OpenSSL < 3.2.0 where the code fails
+         * to correctly source public keys unless explicitly requested
+         * via an expect hint */
+        if (OSSL_STORE_expect(store, OSSL_STORE_INFO_PUBKEY) != 1) {
+            fprintf(stderr, "Failed to expect Public Key File\n");
+            exit(EXIT_FAILURE);
+        }
+    }
+
+    for (info = OSSL_STORE_load(store); info != NULL;
+         info = OSSL_STORE_load(store)) {
+        int type = OSSL_STORE_INFO_get_type(info);
+
+        if (key != NULL) {
+            fprintf(stderr, "Multiple keys matching URI: %s\n", uri);
+            exit(EXIT_FAILURE);
+        }
+
+        switch (type) {
+        case OSSL_STORE_INFO_PUBKEY:
+            key = OSSL_STORE_INFO_get1_PUBKEY(info);
+            break;
+        case OSSL_STORE_INFO_PKEY:
+            key = OSSL_STORE_INFO_get1_PKEY(info);
+            break;
+        }
+        OSSL_STORE_INFO_free(info);
+    }
+
+    if (key == NULL) {
+        fprintf(stderr, "Failed to load key from URI: %s\n", uri);
+        ossl_err_print();
+        exit(EXIT_FAILURE);
+    }
+    OSSL_STORE_close(store);
+
+    return key;
+}
+
+static void test_pkcs1_with_tls_padding(void)
+{
+    EVP_PKEY_CTX *ctx;
+    EVP_PKEY *prikey;
+    EVP_PKEY *pubkey;
+    unsigned char plain[SSL_MAX_MASTER_KEY_LENGTH + 2] = { 0x03, 0x03, 0x01 };
+    unsigned char enc[1024];
+    unsigned char dec[1024];
+    size_t enclen;
+    size_t declen;
+    unsigned int ver = 0x0303;
+    const OSSL_PARAM ver_params[] = {
+        OSSL_PARAM_uint(OSSL_ASYM_CIPHER_PARAM_TLS_CLIENT_VERSION, &ver),
+        OSSL_PARAM_END
+    };
+    int err;
+
+    pubkey = load_key(getenv("PUBURI"));
+
+    ctx = EVP_PKEY_CTX_new_from_pkey(NULL, pubkey, NULL);
+    if (!ctx) {
+        fprintf(stderr, "Failed to init pkey ctx for puburi\n");
+        ossl_err_print();
+        exit(EXIT_FAILURE);
+    }
+    err = EVP_PKEY_encrypt_init(ctx);
+    if (err != 1) {
+        fprintf(stderr, "Failed to init encrypt ctx\n");
+        ossl_err_print();
+        exit(EXIT_FAILURE);
+    }
+    err = EVP_PKEY_CTX_set_rsa_padding(ctx, RSA_PKCS1_PADDING);
```

Comment:
Should this be `RSA_PKCS1_WITH_TLS_PADDING`? Or how does it work when different padding is set for encryption and decryption?

Reply:
No, OpenSSL implement only decryption with that padding, it is a legacy interface, it does not even have a name which is why I had to C code this instead of just using openssl pkeyutl with a rsa_padding_mode option.